compiler/noirc_frontend/src/hir/comptime/interpreter.rs

```diff
@@ -1376,9 +1376,20 @@ impl<'local, 'interner> Interpreter<'local, 'interner> {
     fn evaluate_for(&mut self, for_: HirForStatement) -> IResult<Value> {
         let start_value = self.evaluate(for_.start_range)?;
         let end_value = self.evaluate(for_.end_range)?;
-        let loop_index_type = start_value.get_type();
+        let start_type = start_value.get_type();
+        let end_type = end_value.get_type();
+
+        // Check that start and end have the same type
+        if start_type != end_type {
```

Comment:
We should avoid calling `==` or `!=` on types, it will be incorrect when type variables are involved. You can use `.unify().is_ok()` instead

```suggestion
        if start_type.unify(&end_type).is_ok() {
```